client/rpcproxy/rpcproxy.go

```diff
@@ -0,0 +1,765 @@
+// Package rpcproxy provides a proxy interface for Nomad Servers.  The
+// RpcProxy periodically shuffles which server a Nomad Client communicates
+// with in order to redistribute load across Nomad Servers.  Nomad Servers
+// that fail an RPC request are automatically cycled to the end of the list
+// until the server list is reshuffled.
+//
+// The servers package does not provide any external API guarantees and
+// should be called only by `hashicorp/nomad`.
+package rpcproxy
+
+import (
+	"fmt"
+	"log"
+	"math/rand"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/hashicorp/consul/lib"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+const (
+	// clientRPCJitterFraction determines the amount of jitter added to
+	// clientRPCMinReuseDuration before a connection is expired and a new
+	// connection is established in order to rebalance load across Nomad
+	// servers.  The cluster-wide number of connections per second from
+	// rebalancing is applied after this jitter to ensure the CPU impact
+	// is always finite.  See newRebalanceConnsPerSecPerServer's comment
+	// for additional commentary.
+	//
+	// For example, in a 10K Nomad cluster with 5x servers, this default
+	// averages out to ~13 new connections from rebalancing per server
+	// per second (each connection is reused for 120s to 180s).
+	clientRPCJitterFraction = 2
+
+	// clientRPCMinReuseDuration controls the minimum amount of time RPC
+	// queries are sent over an established connection to a single server
+	clientRPCMinReuseDuration = 600 * time.Second
+
+	// Limit the number of new connections a server receives per second
+	// for connection rebalancing.  This limit caps the load caused by
+	// continual rebalancing efforts when a cluster is in equilibrium.  A
+	// lower value comes at the cost of increased recovery time after a
+	// partition.  This parameter begins to take effect when there are
+	// more than ~48K clients querying 5x servers or at lower server
+	// counts when there is a partition.
+	//
+	// For example, in a 100K Nomad cluster with 5x servers, it will take
+	// ~5min for all servers to rebalance their connections.  If 99,995
+	// agents are in the minority talking to only one server, it will
+	// take ~26min for all servers to rebalance.  A 10K cluster in the
+	// same scenario will take ~2.6min to rebalance.
+	newRebalanceConnsPerSecPerServer = 64
+
+	// rpcAPIMismatchLogRate determines the rate at which log entries are
+	// emitted when the client and server's API versions are mismatched.
+	rpcAPIMismatchLogRate = 3 * time.Hour
+)
+
+// NomadConfigInfo is an interface wrapper around this Nomad Agent's
+// configuration to prevents a cyclic import dependency.
+type NomadConfigInfo interface {
```

Comment:
Potentially remove this interface and create a NewRpcProxy() that just takes these